### PR TITLE
Allow out-of-cluster operation of VPA

### PR DESCRIPTION
**Pull request by JensErat**
Because of how we operate our Kubernetes clusters, we need to run vertical pod autoscaler with an external kubeconfig. Also aligned and factored out redundant configuration code.

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/admission-controller/main.go [hunk 1]
```diff
@@ -58,6 +56,9 @@ var (
 
 	port            = flag.Int("port", 8000, "The port to listen on.")
 	address         = flag.String("address", ":8944", "The address to expose Prometheus metrics.")
+	kubeconfig      = flag.String("kubeconfig", "", "Paths to a kubeconfig. Only required if out-of-cluster.")
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/admission-controller/main.go]: s/Paths/path/

File: vertical-pod-autoscaler/pkg/updater/main.go [hunk 2]
```diff
@@ -72,16 +74,14 @@ func main() {
...
-	if err != nil {
-		klog.Fatalf("Failed to build Kubernetes client : fail to create config: %v", err)
-	}
+	config := common.CreateKubeConfigOrDie(*kubeconfig, float32(*kubeApiQps), int(*kubeApiBurst))
+
 	kubeClient := kube_client.NewForConfigOrDie(config)
 	vpaClient := vpa_clientset.NewForConfigOrDie(config)
 	factory := informers.NewSharedInformerFactory(kubeClient, defaultResyncPeriod)
 	targetSelectorFetcher := target.NewVpaTargetSelectorFetcher(config, kubeClient, factory)
 	var limitRangeCalculator limitrange.LimitRangeCalculator
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/updater/main.go]: nit: You can remove this line now

File: vertical-pod-autoscaler/common/kubeconfig.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package common
+
+import (
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog"
+)
+
+func CreateKubeConfigOrDie(kubeconfig string, kubeApiQps float32, kubeApiBurst int) *rest.Config {
+	//var config *rest.Config
```
Comment (bskiba) [inline vertical-pod-autoscaler/common/kubeconfig.go]: nit: please remove

**Comment by mwielgus** [comment]
@JensErat 
Please address @bskiba comments.

File: vertical-pod-autoscaler/pkg/admission-controller/main.go (hunk 1 above)
Comment (JensErat) [reply vertical-pod-autoscaler/pkg/admission-controller/main.go:59]: I went for a capital `Path` for consistency, I guess you were only after the trailing `s` anyway.

**Comment by JensErat** [comment]
Sorry for the delay, too much other work to do and some vacation in-between. Thank you for the review! Rebased, resolved minors and split up the commit as requested.

**Comment by JensErat** [comment]
Looking at other PRs, it seems master is broken and not just flaky?

    FAIL	k8s.io/autoscaler/cluster-autoscaler/cloudprovider/aws [build failed]

/retest

**Comment by JensErat** [comment]
Seems like #3222 will fix the broken test, will rebase as soon as merged.

**Comment by JensErat** [comment]
@mwielgus @bskiba review comments should be fine now, and tests also work since AWS test is fixed in master.

File: vertical-pod-autoscaler/common/kubeconfig.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (bskiba) [inline vertical-pod-autoscaler/common/kubeconfig.go:2]: nit: s/2018/2020/

File: vertical-pod-autoscaler/pkg/admission-controller/main.go (hunk 1 above)
Comment (bskiba) [reply vertical-pod-autoscaler/pkg/admission-controller/main.go:59]: Hm, I still see the s at the end?

Comment (bskiba) [reply vertical-pod-autoscaler/pkg/admission-controller/main.go:59]: Here and in recommender/main.go

**Comment by bskiba** [comment]
Two more small comments

**Comment by mwielgus** [comment]
Where we are with this PR? It seems that it just needs two nit-fixes and a rebase...

**Comment by mwielgus** [comment]
Inactivity warning. Please resolve the comments and fix conflicts.

**Comment by mwielgus** [comment]
Closing due to inactivity. Feel free to reopen.

**Comment by JensErat** [comment]
/reopen

I'm sorry for the long period of inactivity. I updated most of the minors and will comment on the remaining as soon as prow reopened and github updated my branch.

**Pull request closed without merging**